src/textual/validation.py

```diff
@@ -0,0 +1,505 @@
+"""Framework for validating string values"""
+
+from __future__ import annotations
+
+import math
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Callable, Sequence
+from urllib.parse import urlparse
+
+import rich.repr
+
+from textual._types import Pattern
+
+
+@dataclass
+class ValidationResult:
+    """The result of calling a `Validator.validate` method."""
+
+    failures: Sequence[Failure] = field(default_factory=list)
+    """A list of reasons why the value was invalid. Empty if valid=True"""
+
+    @staticmethod
+    def merge(results: Sequence["ValidationResult"]) -> "ValidationResult":
```

Comment:
Does the Textual dev need this method? I not, could we make it private.

Reply:
They might - if you have a few inputs and you want to know "was everything successful?", and the descriptions for any failures that occurred.